File: utils/SnowFlake.go
package utils

import (
	"fmt"
	"sync"
	"time"
)

// 包内全局变量，保存雪花ID生成的状态
var (
	snowflakeStartTime      int64
	snowflakeSequence       int64
	snowflakeLastTime       int64
	snowflakeSequenceBits   int64
	snowflakeTimestampShift int64
	snowflakeSequenceMask   int64
	snowflakeMutex          sync.Mutex
)

// 初始化雪花ID生成器的参数
func init() {
	// 2025-10-01 00:00:00 UTC 对应的毫秒时间戳
	snowflakeStartTime = int64(1759315200000)
	snowflakeSequenceBits = 5
	snowflakeTimestampShift = snowflakeSequenceBits
	snowflakeSequenceMask = -1 ^ (-1 << snowflakeSequenceBits)
	snowflakeSequence = 0
	snowflakeLastTime = -1
}

// GenerateSnowflakeId 直接生成一个雪花ID
func GenerateSnowflakeId() (int64, error) {
	snowflakeMutex.Lock()
	defer snowflakeMutex.Unlock()

	timestamp := time.Now().UnixMilli()

	// 时钟回退校验
	if timestamp < snowflakeLastTime {
		return 0, fmt.Errorf("clock moved backwards. refuse to generate id for %dms", snowflakeLastTime-timestamp)
	}

	// 同毫秒序列递增，溢出则等待下毫秒
	if snowflakeLastTime == timestamp {
		snowflakeSequence = (snowflakeSequence + 1) & snowflakeSequenceMask
		if snowflakeSequence == 0 {
			// 等待到下一个毫秒
			for timestamp <= snowflakeLastTime {
				timestamp = time.Now().UnixMilli()
			}
		}
	} else {
		snowflakeSequence = 0 // 跨毫秒重置序列
	}

	snowflakeLastTime = timestamp

	// 生成ID：时间戳差左移 + 序列
	return (timestamp-snowflakeStartTime)<<snowflakeTimestampShift | snowflakeSequence, nil
}

File: frontend/src/utils/snowflake.ts
// 雪花ID生成器的状态变量
const snowflakeState = {
    startTime: 1759315200000n, // 2025-10-01 00:00:00 UTC
    sequenceBits: 5n,
    timestampShift: 5n,
    sequenceMask: -1n ^ (-1n << 5n),
    sequence: 0n,
    lastTimestamp: -1n,
    locked: false
};

/**
 * 直接生成一个雪花ID
 * @returns {bigint} 生成的雪花ID
 */
export function generateSnowflakeId(): bigint {
    // 简单的锁机制确保并发安全
    while (snowflakeState.locked) {}
    snowflakeState.locked = true;

    try {
        let timestamp = BigInt(Date.now());

        // 时钟回退校验
        if (timestamp < snowflakeState.lastTimestamp) {
            throw new Error(`clock moved backwards. refuse to generate id for ${Number(snowflakeState.lastTimestamp - timestamp)}ms`);
        }

        // 同毫秒序列递增，溢出则等待下毫秒
        if (snowflakeState.lastTimestamp === timestamp) {
            snowflakeState.sequence = (snowflakeState.sequence + 1n) & snowflakeState.sequenceMask;
            if (snowflakeState.sequence === 0n) {
                // 等待到下一个毫秒
                let newTimestamp = BigInt(Date.now());
                while (newTimestamp <= snowflakeState.lastTimestamp) {
                    newTimestamp = BigInt(Date.now());
                }
                timestamp = newTimestamp;
            }
        } else {
            snowflakeState.sequence = 0n; // 跨毫秒重置序列
        }

        snowflakeState.lastTimestamp = timestamp;

        // 生成ID：时间戳差左移 + 序列
        return (timestamp - snowflakeState.startTime) << snowflakeState.timestampShift | snowflakeState.sequence;
    } finally {
        snowflakeState.locked = false;
    }
}